### ISSUE TITLE
certbot-auto fails on Ubuntu 20.04

### ISSUE THREAD

**Issue by Franky1** [certbot/certbot#7941]
## My operating system is (include version):

Ubuntu 20.04 LTS

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

```shell
wget https://dl.eff.org/certbot-auto
chmod a+x ./certbot-auto
```

## I ran this command and it produced this output:

```shell
sudo ./certbot-auto --install-only --verbose
```

Fails with:

```shell
Reading state information... Done
Note, selecting 'python-is-python2' instead of 'python'
Note, selecting 'python-dev-is-python2' instead of 'python-dev'
Package python-virtualenv is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'python-virtualenv' has no installation candidate
```

**Comment by bmw** [comment]
Yeah we expected this with Ubuntu 20.04's changes to Python. We're planning on deprecating certbot-auto later this year though so we're not going to add support for Ubuntu 20.04 to it.

Our recommended instructions for installing Certbot on Ubuntu 20.04 can be found at https://certbot.eff.org/lets-encrypt/ubuntuother-other.

**Comment by davidfavor** [comment]
https://certbot.eff.org/lets-encrypt/ubuntufocal-other provides bogus instructions.

Following the above instructions installs certbot version 0.40.0-1 which is very old + will fail in many ways over certbot-auto V1.4 (latest).

**Comment by davidfavor** [comment]
Please provide correct installation procedure for Ubuntu Fossa.

Thanks.

**Comment by alissonlauffer** [comment]
> Yeah we expected this with Ubuntu 20.04's changes to Python. We're planning on deprecating certbot-auto later this year though so we're not going to add support for Ubuntu 20.04 to it.
> 
> Our recommended instructions for installing Certbot on Ubuntu 20.04 can be found at https://certbot.eff.org/lets-encrypt/ubuntuother-other.

Why certbot-auto is not ported to Python 3? Since Python 2 is now deprecated.

**Comment by bmw** [comment]
Certbot 0.40.0 is only 6 months old.

With that said, if you'd like a newer version, I'd recommend trying out our [snap instructions](https://certbot.eff.org/lets-encrypt/snap-other). The Certbot snap is still in its beta phase, but if that goes well, it will become our recommended installation instructions for Ubuntu 20.04 in the future.

**Comment by a14m** [comment]
> We're planning on deprecating certbot-auto later this year

Is there any other stable method for installing certbot with a specific version, we use this a lot and have ansible automation workflow based on pinning (new certbot) versions and route53_dns plugin used for wildcard certificates... cannot depend on apt latest version (as it might change unpromptedly)

**Comment by bmw** [comment]
You could [use Docker](https://certbot.eff.org/docs/install.html#running-with-docker).

I think the next best option for you would be to use `pip` directly, however, there are many pitfalls to this approach if you're not familiar with things like virtual environments, conflicts between `pip` and OS package managers, and `pip`'s imperfect dependency resolution so while you can make it work, [this approach is not supported by us](https://certbot.eff.org/docs/install.html#installing-from-source) and we wouldn't be able to help you with any bugs you hit in the process.

**Comment by a14m** [comment]
Docker cannot be an option (due to certain limitations at work)... and due to the fact that the Debian apt packages are fairly old (older than 18 months now) and not having some of the features we actively use (like `pre/post-hooks`)
can we use `snap` for installing? and depend on the fact that snap latest package will work with pip3 `certbot_dns_route53` plugin without getting into version mismatches?

in other words, would the following code fail with version mismatches?
```
sudo snap install --classic certbot
sudo pip install certbot_dns_route53
sudo certbot certonly  -d {{ site_name }} -m {{ email }} --dns-route53 --agree-tos --non-interactive
```

**Comment by bmw** [comment]
For what it's worth, unless you're on Debian 8 Jessie or older, the Debian packages should have pre/post hooks. The [hook directories were added in Certbot 0.19.0](https://github.com/certbot/certbot/blob/master/certbot/CHANGELOG.md#0190---2017-10-04) and the command line flags themselves were added long before that.

You cannot use the snaps with packages from `pip`. We're planning on announcing the beta phase of snaps for our DNS plugins in the next week or so, but the documentation on how to use them is not ready yet. You can track that at https://github.com/certbot/certbot/issues/8142 if you want. You could maybe continue to use `certbot-auto` until that is ready.

I hope that helps, but if you'd like more help with this, I recommend making a post at https://community.letsencrypt.org/. There is a much larger community of people there familiar with Certbot that should be able to help you.

**Comment by fake-name** [comment]
> We're planning on deprecating certbot-auto later this year though so we're not going to add support for Ubuntu 20.04 to it.

This is deeply, **deeply** disappointing. Snapd and docker both horrible solutions.

certbot-auto also works fine on 20.04 by simply patching it to use python3 instead of python2:

 ```

client@lolbox:/opt/letsencrypt$ git diff | cat
diff --git a/certbot-auto b/certbot-auto
index 2a0cda9b3..2d62b78ef 100755
--- a/certbot-auto
+++ b/certbot-auto
@@ -53,7 +53,11 @@ Help for certbot itself cannot be provided until it is installed.
   -q, --quiet                               provide only update/error output;
                                             implies --non-interactive

+
 All arguments are accepted and forwarded to the Certbot client when run."
+
+export USE_PYTHON_3=1
+USE_PYTHON_3=1
 export CERTBOT_AUTO="$0"

 for arg in "$@" ; do
@@ -268,21 +272,13 @@ DeterminePythonVersion() {
   # Arguments: "NOCRASH" if we shouldn't crash if we don't find a good python
   #
   # If no Python is found, PYVER is set to 0.
-  if [ "$USE_PYTHON_3" = 1 ]; then
-    MIN_PYVER=$MIN_PYVER3
-    MIN_PYTHON_VERSION=$MIN_PYTHON_3_VERSION
-    for LE_PYTHON in "$LE_PYTHON" python3; do
-      # Break (while keeping the LE_PYTHON value) if found.
-      $EXISTS "$LE_PYTHON" > /dev/null && break
-    done
-  else
-    MIN_PYVER=$MIN_PYVER2
-    MIN_PYTHON_VERSION=$MIN_PYTHON_2_VERSION
-    for LE_PYTHON in "$LE_PYTHON" python2.7 python27 python2 python; do
-      # Break (while keeping the LE_PYTHON value) if found.
-      $EXISTS "$LE_PYTHON" > /dev/null && break
-    done
-  fi
+  MIN_PYVER=$MIN_PYVER3
+  MIN_PYTHON_VERSION=$MIN_PYTHON_3_VERSION
+  for LE_PYTHON in "$LE_PYTHON" python3; do
+    # Break (while keeping the LE_PYTHON value) if found.
+    $EXISTS "$LE_PYTHON" > /dev/null && break
+  done
+
   if [ "$?" != "0" ]; then
     if [ "$1" != "NOCRASH" ]; then
       error "Cannot find any Pythons; please install one!"
@@ -344,7 +340,7 @@ BootstrapDebCommon() {
   fi

   if apt-cache show python-virtualenv > /dev/null 2>&1; then
-    virtualenv="$virtualenv python-virtualenv"
+    virtualenv="$virtualenv python3-venv"
   fi

   augeas_pkg="libaugeas0 augeas-lenses"
@@ -354,8 +350,8 @@ BootstrapDebCommon() {
   fi

   apt-get install $QUIET_FLAG $YES_FLAG --no-install-recommends \
-    python \
-    python-dev \
+    python3 \
+    python3-dev \
     $virtualenv \
     gcc \
     $augeas_pkg \

```

Please don't drop this tool that works fine!

**Comment by saghul** [comment]
I got it working on Debian 10 just by manually installing `python3-venv` and then running: `USE_PYTHON_3=1 certbot-auto --noninteractive --install-only --no-bootstrap` so no patches were necessary, just skipping the bootstrap should do.

**Comment by smacz42** [comment]
Had to revise the title of that issue - this is actually now affecting Ubuntu 18.04.5 servers.

**Comment by saghul** [comment]
Yes, 1.10 dropped support for Debian based systems. Right now things are in a pretty uncomfortable situation for those of us running Debian, specially on containers, since the proposed solution of using snap just doesn't work out of the box.

What I mentioned in https://github.com/certbot/certbot/issues/7941#issuecomment-734812632 works if you add `--no-selfupdate` (I think the option was called that way) but for some odd reason doesn't work on Debian 10 armhf.

I ran out of talent so I ended up migrating to acme.sh as I couldn't find a solution.

**Comment by bmw** [comment]
If acme.sh works better for your use case, by all means use it. The important thing is that you're able to secure your site.

If you still want to use Certbot though, my post at https://community.letsencrypt.org/t/certbot-auto-no-longer-works-on-debian-based-systems/139702/7 and in particular the alternate installation instructions documented at https://certbot.eff.org/docs/install.html may help you. (certbot-auto is still documented there but that will be removed soon.)
